// localhost.daplie.me-certificates/lib/index.js
'use strict';

// Assumptions:
//   you have a project with a web app such as ~/my-project/lib/server.js
//   you will place the certs in ~/my-project/certs/
//
// git clone https://git.daplie.com/Daplie/localhost.daplie.com-certificates.git ../certs

var fs = require('fs');
var path = require('path');
var tls = require('tls');

var defaultServername = 'localhost.daplie.me';
var servernames = fs.readdirSync(path.join(__dirname, '..', 'certs'));
var secureContexts = {};
var certsMap = {};

//
// SSL Certificates
//
var options = {
  key: fs.readFileSync(path.join(__dirname, '..', 'privkey.pem'))
, cert: fs.readFileSync(path.join(__dirname, '..', 'fullchain.pem'))
, requestCert: false
, rejectUnauthorized: true

  // If you need to use SNICallback you should be using io.js >= 1.x (possibly node >= 0.12)
, SNICallback: function (servername, cb) {
    // null means all requests will default to this certificate
    // var secureContext = null;
    // Instead we explicitly set the context to this one (for example's sake)

    if ('function' === typeof cb) {
      cb(null, secureContexts[servername] || module.exports.getTlsContext(servername, this) || secureContexts[defaultServername]);
    } else {
      console.error('your version of node is too old to properly handle SNICallback');
      return secureContexts[servername] || module.exports.getTlsContext(servername, this) || secureContexts[defaultServername];
    }
  }
};

module.exports = options.options = options;
module.exports.create = function (opts) {
  opts = opts || {};

  var newObj = {};
  Object.keys(opts).forEach(function (key) {
    newObj[key] = opts[key];
  });

  return module.exports.merge(newObj);
};
module.exports.merge = function (opts) {
  opts = opts || {};

  Object.keys(options).forEach(function (key) {
    if (!(key in opts)) {
      opts[key] = options[key];
    }
  });

  // prevent circular refs (until v2.0.0 when this can be removed)
  delete opts.merge;
  delete opts.create;
  delete opts.options;
  return opts;
};

module.exports.mergeTlsOptions = function (servername, opts) {
  opts = opts || {};

  if (-1 === servernames.indexOf(servername)) {
    return null;
  }

  if (!certsMap[servername]) {
    certsMap[servername] = {
      servername: servername
    , key: fs.readFileSync(path.join(__dirname, '..', 'certs', servername, 'privkey.pem'))
    , cert: fs.readFileSync(path.join(__dirname, '..', 'certs', servername, 'fullchain.pem'))
    };
  }

  opts.key = certsMap[servername].key;
  opts.cert = certsMap[servername].cert;
  return opts;
};

module.exports.getTlsContext = function (servername, opts) {
  opts = opts || {};

  if (!secureContexts[servername]) {
    // TODO clone?
    var fullOpts = module.exports.mergeTlsOptions(servername, opts);
    if (fullOpts) {
      secureContexts[servername] = tls.createSecureContext(opts);
    }
  }

  return secureContexts[servername];
};

module.exports.getAllSecureContexts = function (opts) {
  servernames.forEach(function (servername) {
    module.exports.getTlsContext(servername, opts);
  });

  return secureContexts;
};

module.exports.getTlsContext('localhost.daplie.me', {});
// var server = https.createServer(options || require('localhost.daplie.com-certificates'));
// server.on('request', function (req, res) { res.end('hello'); });
// server.listen(443, function () { console.log('<https://localhost.daplie.com>'); });
